File: app/src/main/java/com/alex/materialdiary/sys/adapters/RecycleAdapterMarks.java
package com.alex.materialdiary.sys.adapters;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.alex.materialdiary.R;
import com.alex.materialdiary.sys.common.models.period_marks.Mark;
import com.alex.materialdiary.sys.common.models.period_marks.PeriodMarksData;
import com.google.android.material.card.MaterialCardView;

import java.util.Calendar;
import java.util.List;
import java.util.Objects;

import javax.security.auth.callback.Callback;

public class RecycleAdapterMarks extends RecyclerView.Adapter<RecycleAdapterMarks.ViewHolder>{

    private final LayoutInflater inflater;
    private final List<Mark> periods;
    private final List<com.alex.materialdiary.sys.common.models.marks.Mark> diffs;

    public RecycleAdapterMarks(Context context, List<Mark> periods, List<com.alex.materialdiary.sys.common.models.marks.Mark> diffs) {
        this.inflater = LayoutInflater.from(context);
        this.periods = periods;
        this.diffs = diffs;
    }

    @NonNull
    @Override
    public RecycleAdapterMarks.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.mark_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Mark mark = periods.get(position);
        MaterialCardView itemView = (MaterialCardView) holder.itemView;
        itemView.setStrokeColor(inflater.getContext().getResources().getColor(R.color.gray));
        if (diffs.contains(new com.alex.materialdiary.sys.common.models.marks.Mark(mark.getValue(), mark.getDate()))){
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    itemView.setStrokeColor(Color.RED);
            }
        }
        Context c = inflater.getContext();
        Object year = Calendar.getInstance().get(Calendar.YEAR);
        String yearr = String.valueOf(Calendar.getInstance().get(Calendar.YEAR));
        if(mark.getDate() != null){
            String text = mark.getDate();
            try {
                String pr_year = String.valueOf(((int)year) - 1);
                text = text.replace("." + pr_year, "");
            } catch (Exception ignored) {
            }
            holder.date.setText(text.replace("." + yearr, ""));
        }
        holder.mark.setText(String.valueOf(mark.getValue()));
        //holder.itemView.setOnClickListener(v -> Toast.makeText(c, String.valueOf(mark.get), Toast.LENGTH_SHORT).show());
        if (mark.getValue() == 5) holder.mark.setTextColor(c.getResources().getColor(R.color.five));
        else if (mark.getValue() == 4) holder.mark.setTextColor(c.getResources().getColor(R.color.four));
        else if (mark.getValue() == 3) holder.mark.setTextColor(c.getResources().getColor(R.color.three));
        else if (mark.getValue() == 2) holder.mark.setTextColor(c.getResources().getColor(R.color.two));
        else if (mark.getValue() == 1) holder.mark.setTextColor(c.getResources().getColor(R.color.one));
    }

    @Override
    public int getItemCount() {
        return periods.size();
    }
    public static class ViewHolder extends RecyclerView.ViewHolder {
        final TextView date;
        final TextView mark;

        ViewHolder(View view){
            super(view);
            date = view.findViewById(R.id.date_of_mark);
            mark = view.findViewById(R.id.mark);
        }
    }
}

File: app/src/main/java/com/alex/materialdiary/sys/adapters/RecycleAdapterMarks.kt
package com.alex.materialdiary.sys.adapters

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.alex.materialdiary.R
import com.google.android.material.card.MaterialCardView
import android.os.Build
import android.util.TypedValue
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import androidx.preference.Preference
import androidx.preference.PreferenceManager
import com.alex.materialdiary.sys.net.models.period_marks.Mark
import com.google.android.material.badge.BadgeDrawable
import com.google.android.material.badge.BadgeUtils
import java.lang.Exception
import java.util.*
import kotlin.coroutines.coroutineContext

class RecycleAdapterMarks(
    context: Context?,
    periods: List<Mark>,
    diffs: List<com.alex.materialdiary.sys.net.models.marks.Mark>
) : RecyclerView.Adapter<RecycleAdapterMarks.ViewHolder>() {
    private val inflater: LayoutInflater
    private val periods: List<Mark>
    private val diffs: List<com.alex.materialdiary.sys.net.models.marks.Mark>
    val useBadge: Boolean = PreferenceManager.getDefaultSharedPreferences(context).getBoolean("useBadge", false)
    val newColor = Color.parseColor("#FF6750A4")
    init {
        inflater = LayoutInflater.from(context)
        this.periods = periods
        this.diffs = diffs
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = inflater.inflate(R.layout.mark_item, parent, false)
        return ViewHolder(view)
    }
    override fun  onBindViewHolder(holder: ViewHolder, position: Int) {
        val mark = periods[position]
        holder.card.cardElevation = 4F
        holder.card.strokeColor = Color.TRANSPARENT
        holder.detachBadge()
        if (diffs.contains(
                com.alex.materialdiary.sys.net.models.marks.Mark(
                    mark.value,
                    mark.date
                )
            )
        ) {
            //itemView.setBackgroundColor(inflater.context.resources.getColor(R.color.new_mark))
            //itemView.setCardBackgroundColor(inflater.context.resources.getColor(R.color.new_mark))
            //holder.card.strokeColor = inflater.context.resources.getColor(R.color.two)
            //holder.badgeDrawable.isVisible = true
            //holder.card.cardElevation = 0F
            if (useBadge) holder.attachBadge()
            else {
                holder.card.strokeColor = newColor
                holder.card.cardElevation = 0F
            }
        }
        val c = inflater.context
        val year: Any = Calendar.getInstance()[Calendar.YEAR]
        val yearr = Calendar.getInstance()[Calendar.YEAR].toString()
        var text = mark.date
        try {
            val pr_year = (year as Int - 1).toString()
            text = text.replace(".$pr_year", "")
        } catch (_: Exception) {
        }
        holder.date.text = text.replace(".$yearr", "")
        holder.mark.text = mark.value.toString()
        if (mark.value in 1..5){
            holder.mark.setTextColor(Color.WHITE)
            holder.date.setTextColor(Color.WHITE)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            holder.view.tooltipText = mark.typeName
            if (mark.typeName == null) holder.view.tooltipText = "Тип не указан"
        }
        holder.view.setOnClickListener { obj: View -> obj.performLongClick() }
        if (mark.value == 5) holder.card.backgroundTintList = ColorStateList.valueOf(c.resources.getColor(R.color.five_tint))
        else if (mark.value == 4) holder.card.backgroundTintList = ColorStateList.valueOf(c.resources.getColor(R.color.four_tint))
        else if (mark.value == 3) holder.card.backgroundTintList = ColorStateList.valueOf(c.resources.getColor(R.color.three_tint))
        else if (mark.value == 2) holder.card.backgroundTintList = ColorStateList.valueOf(c.resources.getColor(R.color.two_tint))
        else if (mark.value == 1) holder.card.backgroundTintList = ColorStateList.valueOf(c.resources.getColor(R.color.one_tint))
    }

    override fun getItemCount(): Int {
        return periods.size
    }
    @androidx.annotation.OptIn(com.google.android.material.badge.ExperimentalBadgeUtils::class)
    class ViewHolder internal constructor(val view: View) : RecyclerView.ViewHolder(view) {
        val date: TextView
        val mark: TextView
        val card: MaterialCardView
        val badgeDrawable: BadgeDrawable
        fun Context.toPx(dp: Int): Float = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            dp.toFloat(),
            resources.displayMetrics)
        init {
            date = view.findViewById(R.id.date_of_mark)
            mark = view.findViewById(R.id.mark)
            card = view.findViewById(R.id.anchor_for_badge)
            badgeDrawable = BadgeDrawable.create(view.context)
            //badgeDrawable.badgeGravity = BadgeDrawable.T
            badgeDrawable.text = " "
            //badgeDrawable.isVisible = false
            //badgeDrawable.clearText()
            //badgeDrawable.horizontalOffset = view.context.toPx(45).toInt()
            card.addOnLayoutChangeListener { _, _, _, _, _, _, _, _, _ ->
                badgeDrawable.updateBadgeCoordinates(card, view as FrameLayout)
            }
        }
        fun attachBadge(){
            BadgeUtils.attachBadgeDrawable(badgeDrawable, card, view as FrameLayout)
        }
        fun detachBadge(){
            BadgeUtils.detachBadgeDrawable(badgeDrawable, card)
        }
    }
}